rclcpp/src/rclcpp/executor.cpp

```diff
@@ -232,16 +232,24 @@ Executor::spin_some(std::chrono::nanoseconds max_duration)
     throw std::runtime_error("spin_some() called while already spinning");
   }
   RCLCPP_SCOPE_EXIT(this->spinning.store(false); );
-  // non-blocking call to pre-load all available work
-  wait_for_work(std::chrono::milliseconds::zero());
-  while (spinning.load() && max_duration_not_elapsed()) {
-    AnyExecutable any_exec;
-    if (get_next_ready_executable(any_exec)) {
-      execute_any_executable(any_exec);
-    } else {
-      break;
+  do {
```

Comment:
@ivanpauno one suggestion:

```c++
bool work_available = false;
AnyExecutable any_exec;
while (spinning.load() && max_duration_not_elapsed()) {
  if (!work_available) {
    wait_for_work(std::chrono::milliseconds::zero());
  }
  if (get_next_ready_executable(any_exec)) {
      execute_any_executable(any_exec);
      work_available = true;
  } else {
     if (!work_available) {
        break;
     }
     work_available = false;
  }
}
```

Reply:
that's wrong, because it'll wait again until executing one executable.
That's what https://github.com/ros2/rclcpp/pull/844 fixed.

Reply:
I think this loop structure is a bit cleaner, so probably would go this way if possible.

Reply:
That was almost perfect.
I added an extra check, see https://github.com/ros2/rclcpp/pull/1156/commits/f4dc1768650d14a392f563161eaf9aafffb05cd8.

Reply:
Good catch.